R/is_text_file.R

```diff
@@ -0,0 +1,46 @@
+#' Guess if a file is a text file for a vector of filenames or filepaths
+#'
+#' This function will guess if a file is a text file for a vector of filenames
+#' or filepaths and return `TRUE` or `FALSE` for each. It primarily uses the
+#' system's `file -b --mime-type` to guess the content type, but if `file` is
+#' not found on the system's path (e.g. on Windows without Rtools installed)
+#' and/or if the `find` path is not executable it will fall back on using
+#' `wand::get_content_type()`, which is R-only and cross platform compatible
+#' but less robust.
+#'
+#' @param filepaths A character vector
+#'
+#' @return A logical vector the same length as `filepaths`.
+#' @export
+is_text_file <- function(filepaths) {
+  if (is.data.frame(filepaths) && identical(length(filepaths), 1L)) {
+    filepaths <- filepaths[[1L]]
+  }
+
+  filepaths <- enc2utf8(as.character(filepaths))
+  filepaths <- fs::path_abs(fs::path_expand(filepaths))
+
+  mime_type <- function(filepath) {
+    file_command <- Sys.which("file")
+    if (!is_file_accessible(filepath)) {
+      NA_character_
+    } else if (fs::file_access(file_command, mode = "execute")) {
+      system2(
+        command = "file",
+        args = c("-b", "--mime-type", shQuote(filepath)),
+        stdout = TRUE
+      )
+    } else {
```

Comment:
Can we change this to also `require("wand")`, and then throw a warning if it isn't installed? Maybe it should error if it can't check, but won't we see the error in the file failing to open anyways?

Reply:
I prefer to have these low level functions just work and not throw a bunch of errors/warnings/messages over bad input. I could add one more `require(wand)` here and then add one more `else` that just returns `NA_character`. 

The intent was for all of these `is_*()` style functions to **always* return `TRUE` or `FALSE` for anything you pass to it.... if it can't definitively guarantee that something "is" what it is supposed to be, it should return `FALSE`, regardless of why it could not verify it.

Reply:
but.... do you mean use `require(wand)` here and then remove `{wand}` as a dependency? I could do that, but.... honestly `{wand}` is a rather obscure package, so if we don't import it, there's an extremely high probability it won't be installed, so.... might as well just remove it completely.... thoughts?

Reply:
actually, I'm rather opposed to that now that I think about it because the only likely way a Windows user will have access to the `file` command is if Rtools is installed, which is not too uncommon because it's need to build packages from source, but I suspect there's a substantial chunk of R users that don't have it installed since it's a separate download and install and requires a bit of setup.

Reply:
> remove `{wand}` as a dependency

I was thinking it could be a `suggests`.

Reply:
Yeah, I get it. But since `Suggests` packages are not installed by default with `base::install.packages()`, I have to assume most/many Windows users will not have either of the facilities to make this function work (`files` with Rtools, or `{wand}`), and therefore I assume for most/many Windows users this function will return `FALSE` for anything you give it, making it rather useless.

Maybe I should merge into this repo some of the file type guessing based on file extension? e.g. if none of the other tools are available, assume any path that ends in `.csv`, `.tsv`, `.txt`, `.R`, `.Rmd`, `.md`, `.yml`, `.yaml`, etc. is a text file?

Reply:
I've completely reimplemented this function by checking for nul bytes in the beginning of the file, which has no package dependencies and no dependency on the system's `file` command.